From 8ea520148a7ef8212b460c2e52719ecf6a34940d Mon Sep 17 00:00:00 2001
From: Paul Thomson <paulthomson@google.com>
Date: Thu, 16 Apr 2026 15:39:31 +0100
Subject: [PATCH 2/5] Various analysis API changes

Mostly changing visibility, but also some other changes. Could split.
---
 .../analysis/api/descriptors/CliFe10AnalysisFacade.kt       | 6 ++++--
 .../fe10/base/DummyKtFe10ReferenceResolutionHelper.kt       | 3 ++-
 .../base/projectStructure/PluginStructureProvider.kt        | 1 -
 .../api/standalone/StandaloneAnalysisAPISessionBuilder.kt   | 4 ++++
 .../light/classes/symbol/SymbolKotlinAsJavaSupport.kt       | 4 ++--
 .../classes/symbol/annotations/symbolAnnotationsUtils.kt    | 3 ++-
 6 files changed, 14 insertions(+), 7 deletions(-)

diff --git a/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/analysis/api/descriptors/CliFe10AnalysisFacade.kt b/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/analysis/api/descriptors/CliFe10AnalysisFacade.kt
index 1dfbffba1173..82ef6a8bb467 100644
--- a/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/analysis/api/descriptors/CliFe10AnalysisFacade.kt
+++ b/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/analysis/api/descriptors/CliFe10AnalysisFacade.kt
@@ -35,7 +35,8 @@ import org.jetbrains.kotlin.resolve.lazy.ResolveSession
 import org.jetbrains.kotlin.types.checker.KotlinTypeRefiner
 import org.jetbrains.kotlin.util.CancellationChecker
 
-internal class CliFe10AnalysisFacade : Fe10AnalysisFacade {
+// Expose to be registered at FE10UastEnvironment
+class CliFe10AnalysisFacade : Fe10AnalysisFacade {
     override fun getAnalysisContext(element: KtElement, token: KaLifetimeToken): Fe10AnalysisContext {
         val handler = getHandler(element)
         return getAnalysisContext(handler, token)
@@ -78,7 +79,8 @@ internal class CliFe10AnalysisFacade : Fe10AnalysisFacade {
         this ?: error("Resolution is not performed")
 }
 
-internal class KaFe10AnalysisHandlerExtension(
+// Expose to be registered at FE10UastEnvironment
+class KaFe10AnalysisHandlerExtension(
     private val useSiteModule: KaSourceModule? = null
 ) : AnalysisHandlerExtension {
     internal companion object {
diff --git a/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/references/fe10/base/DummyKtFe10ReferenceResolutionHelper.kt b/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/references/fe10/base/DummyKtFe10ReferenceResolutionHelper.kt
index 86b221141436..aeec63e7774d 100644
--- a/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/references/fe10/base/DummyKtFe10ReferenceResolutionHelper.kt
+++ b/analysis/analysis-api-fe10/src/org/jetbrains/kotlin/references/fe10/base/DummyKtFe10ReferenceResolutionHelper.kt
@@ -22,7 +22,8 @@ import org.jetbrains.kotlin.resolve.BindingContext
  * The actual implementation in IDE, [KtFe10ReferenceResolutionHelperImpl], is not always available, e.g., in Android Lint CLI.
  * Other CLI clients that use reference resolution via PSI utils may need to register this dummy implementation too.
  */
-internal object DummyKtFe10ReferenceResolutionHelper : KtFe10ReferenceResolutionHelper {
+// Expose to be registered at FE10UastEnvironment
+object DummyKtFe10ReferenceResolutionHelper : KtFe10ReferenceResolutionHelper {
     override fun isInProjectOrLibSource(element: PsiElement, includeScriptsOutsideSourceRoots: Boolean): Boolean = false
 
     override fun resolveImportReference(file: KtFile, fqName: FqName): Collection<DeclarationDescriptor> = emptyList()
diff --git a/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/projectStructure/PluginStructureProvider.kt b/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/projectStructure/PluginStructureProvider.kt
index c4da60b72e0d..7d1b98c29ffe 100644
--- a/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/projectStructure/PluginStructureProvider.kt
+++ b/analysis/analysis-api-standalone/analysis-api-standalone-base/src/org/jetbrains/kotlin/analysis/api/standalone/base/projectStructure/PluginStructureProvider.kt
@@ -113,7 +113,6 @@ object PluginStructureProvider {
 
         val listenersMap = ConcurrentHashMap<String, MutableList<ListenerDescriptor>>()
         for (listenerDescriptor in listenerDescriptors) {
-            listenerDescriptor.pluginDescriptor = fakePluginDescriptor
             listenersMap.computeIfAbsent(listenerDescriptor.topicClassName) { SmartList() }.add(listenerDescriptor)
         }
 
diff --git a/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/api/standalone/StandaloneAnalysisAPISessionBuilder.kt b/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/api/standalone/StandaloneAnalysisAPISessionBuilder.kt
index 4a9bea987701..98d0a9b6e771 100644
--- a/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/api/standalone/StandaloneAnalysisAPISessionBuilder.kt
+++ b/analysis/analysis-api-standalone/src/org/jetbrains/kotlin/analysis/api/standalone/StandaloneAnalysisAPISessionBuilder.kt
@@ -61,6 +61,7 @@ import kotlin.contracts.contract
 public class StandaloneAnalysisAPISessionBuilder(
     projectDisposable: Disposable,
     unitTestMode: Boolean,
+    compilerConfiguration: CompilerConfiguration = CompilerConfiguration(),
 ) {
     init {
         // We depend on swing (indirectly through PSI or something), so we want to declare headless mode,
@@ -75,6 +76,7 @@ public class StandaloneAnalysisAPISessionBuilder(
         StandaloneProjectFactory.createProjectEnvironment(
             projectDisposable,
             KotlinCoreApplicationEnvironmentMode.fromUnitTestModeFlag(unitTestMode),
+            compilerConfiguration,
         )
 
     private val serviceRegistrars = listOf(FirStandaloneServiceRegistrar, StandaloneSessionServiceRegistrar)
@@ -258,6 +260,7 @@ internal object StandaloneSessionServiceRegistrar : AnalysisApiSimpleServiceRegi
 public inline fun buildStandaloneAnalysisAPISession(
     projectDisposable: Disposable = Disposer.newDisposable("StandaloneAnalysisAPISession.project"),
     unitTestMode: Boolean = false,
+    compilerConfiguration: CompilerConfiguration = CompilerConfiguration(),
     init: StandaloneAnalysisAPISessionBuilder.() -> Unit,
 ): StandaloneAnalysisAPISession {
     contract {
@@ -266,6 +269,7 @@ public inline fun buildStandaloneAnalysisAPISession(
     return StandaloneAnalysisAPISessionBuilder(
         projectDisposable,
         unitTestMode,
+        compilerConfiguration,
     ).apply(init).build()
 }
 
diff --git a/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/SymbolKotlinAsJavaSupport.kt b/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/SymbolKotlinAsJavaSupport.kt
index 45ca7817b185..d0f9cc9d0d91 100644
--- a/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/SymbolKotlinAsJavaSupport.kt
+++ b/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/SymbolKotlinAsJavaSupport.kt
@@ -93,8 +93,8 @@ fun <T> withMultiplatformLightClassSupport(block: () -> T): T {
         return block()
     }
 
-    require(ApplicationManager.getApplication().isReadAccessAllowed) { "The method can only run inside a read action" }
-    require(!ApplicationManager.getApplication().isWriteAccessAllowed) { "The method cannot be run inside a write action" }
+    // require(ApplicationManager.getApplication().isReadAccessAllowed) { "The method can only run inside a read action" }
+    // require(!ApplicationManager.getApplication().isWriteAccessAllowed) { "The method cannot be run inside a write action" }
 
     try {
         KMP_CACHE.set(WeakHashMap())
diff --git a/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/annotations/symbolAnnotationsUtils.kt b/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/annotations/symbolAnnotationsUtils.kt
index c5387906acde..b6d132bb3dad 100644
--- a/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/annotations/symbolAnnotationsUtils.kt
+++ b/analysis/symbol-light-classes/src/org/jetbrains/kotlin/light/classes/symbol/annotations/symbolAnnotationsUtils.kt
@@ -36,7 +36,8 @@ internal fun KaAnnotatedSymbol.hasJvmSyntheticAnnotation(): Boolean {
     return JVM_SYNTHETIC_ANNOTATION_CLASS_ID in annotations
 }
 
-internal fun KaAnnotatedSymbol.getJvmNameFromAnnotation(): String? {
+// Expose to use at KotlinStaticPsiDeclarationFromBinaryModuleProvider
+fun KaAnnotatedSymbol.getJvmNameFromAnnotation(): String? {
     return stringArgumentFromAnnotation(JvmStandardClassIds.Annotations.JvmName)
 }
 
-- 
2.54.0.rc1.513.gad8abe7a5a-goog

